### DISCUSSION TITLE
Best practices for handling static values that don't need to be "real" data (composite types?)

### DISCUSSION THREAD

**Discussion by bennettdams** [prisma/prisma#14302] [Q&A] [answered]
I would like to ask how _you_ handle static values. I feel like I'm missing something for this topic, maybe there are any best practices around this?

---

Imagine a schema model like so:
- A user can have multiple favorite songs (an arbitrary number of songs).
- He provides links (arbitrary URLs) to his favorite songs on his profile.
- We (our app) only allow certain music platforms for the song link (like Spotify or SoundCloud).
- Each music platform has some additional info, like a name or a logo.

Here is how such schema could look like:

```prisma
model User {
  userId           String        @id @default(cuid())
  createdAt        DateTime      @default(now())
  updatedAt        DateTime      @updatedAt
  username         String
  // song links relation
  favoriteSongs    SongLink[]    @relation("User_SongLink")
}

model SongLink {
  songLinkId        String        @id @default(cuid())
  // this is the arbitrary URL from user input on his profile page
  address           String
  // user relation
  User              User          @relation("User_SongLink", fields: [userId], references: [userId])
  userId            String
  // music platform relation
  MusicPlatform     MusicPlatform @relation("SongLink_MusicPlatform", fields: [musicPlatformId], references: [musicPlatformId])
  musicPlatformId   String
}

model MusicPlatform {
  musicPlatformId    String         @id
  name               String
  logoId             String
  // song links relation
  songLinks          SongLink[]     @relation("SongLink_MusicPlatform")
}
```

![image](https://user-images.githubusercontent.com/29319414/178990395-b002a477-438c-4eb8-8a52-33c0522b0071.png)

Here is my question: How do _you_  create such music platforms and use it in your app code?

---

Here are two possible ways to do this that have different questions attached to them:

1. Create the music platforms as "real" data in the database.
1.1 Should they have a unique string ID (e.g. `spotify`) or an automatically-generated ID?
1.2 What should be done when we remove a music platform from our app?
1.3 We ALWAYS need values for these (as our app doesn't make sense without music platforms), even when we create a fresh database, so how to initialize the data? Seed script? Or as part of a migration? 

2. Create the music platforms only as TypeScript variable.
2.1 This variable lives in some central file and can be used for the value (e.g. to get the `name` of a music platform), something like:

    ```ts
    const musicPlatforms = [
      { musicPlatformId: "spotify", name: "Spotify" },
      { musicPlatformId: "soundcloud", name: "SoundCloud" },
    ] as const;
    ```

    This can be reused throughout the app code, but the database is not aware of it at this point. For the database, the `musicPlatformId` would be an arbitrary string, so one would need to build linking (and validating) a `SongLink` to a `MusicPlatform` into his app code. Maybe one could add this as an enum and reference it in the `SongLink` model?

---

For both variants, from an app code perspective, this is more unpleasant to deal with. There are multiple app functionality scenarios that we need to take care of, like:
  - The user found a new song and adds a new song link -> we need to have an API for creating song links
  - The user found a remix for a song that he already linked. So he changes the song link address -> we need to have an API for updating specific song links
  
#### ..but this could be extremely simplified:

In an ideal world, the user's `favoriteSongs` are is array of objects (see `SongLink`), where the `musicPlatformId` is checked for existence/correctness by the DB. This would allow us to only needing **one API** that receives an array of objects (a `SongLink` array), that we just use to overwrite a user's `favoriteSongs` field.

I think something like this is possible with [composite types of the MongoDB data model](https://www.prisma.io/docs/concepts/components/prisma-schema/data-model#defining-composite-types), but how do you deal with this in e.g. Postgres?

**Comment by austincrim** [comment, marked as answer]
Hey @bennettdams, thanks for the detailed question!

I would store the music platforms in the database as their own table, like you have modeled in the Prisma schema above.

> 1.1 Should they have a unique string ID (e.g. spotify) or an automatically-generated ID?

This will be a matter of preference. I don't see any tradeoffs one way or the other.

> 1.2 What should be done when we remove a music platform from our app?

It depends on how you as the developer want the application to behave. You could store some kind of `supported` or `valid` flag on the `MusicPlatform` model to tell what's valid while still keeping all the data in the table.

> 1.3 We ALWAYS need values for these (as our app doesn't make sense without music platforms), even when we create a fresh database, so how to initialize the data? Seed script? Or as part of a migration?

I would use Prisma's [integrated seeding](https://www.prisma.io/docs/guides/database/seed-database) for this.

**Reply by bennettdams** [reply]
Thanks for your reply! What do you say about the disadvantages I wrote down, namely creating dedicated APIs, just to create such "song links" that can be connected to an "music platform"? In other words, having the "music platform" as a real model that is used in a "song link", you always has to connect them like such. I feel like this is an "unnecessary" overhead, because I already know the music platform IDs anyway, so I don't _need_ to have them as such in the database.

Quoting myself here:

> In an ideal world, the user's favoriteSongs are is array of objects (see SongLink), where the musicPlatformId is checked for existence/correctness by the DB. This would allow us to only needing one API that receives an array of objects (a SongLink array), that we just use to overwrite a user's favoriteSongs field.

 I feel like there has to be an easier way that I'm just not aware of?

**Reply by austincrim** [reply]
I like the idea of using the database as a central place to control the valid music platforms instead of keeping that logic in the app code.

You could keep a plain string (or enum) `musicPlatform` field on the `Song` model, but that requires a little extra validation in your code. Ultimately, both ways should work and it comes down to personal preference.

> I think something like this is possible with [composite types of the MongoDB data model](https://www.prisma.io/docs/concepts/components/prisma-schema/data-model#defining-composite-types), but how do you deal with this in e.g. Postgres?

We intend on bring type JSON or composite types to other database providers as well, but we haven't prioritized it quite yet.

**Reply by bennettdams** [reply]
Hey @austincrim,
I think I was asking two questions at the same time, which made it hard to follow.

For my "optimal" solution (_... the user's favoriteSongs are is array of objects (see SongLink), where the musicPlatformId is checked for existence/correctness by the DB_), I'm now using `upsert` in a transaction, similar like shown in [this feature request](https://github.com/prisma/prisma/discussions/7547). This works fine!

For the other topic, which is the static values, I'm still not sure. I'm using Prisma's seed already, but is this really a good place for such values? I feel like this should be part of a migration, so it will always be filled in the DB. Makes it also easier to add/remove static values. 

I am using the static values now as IDs for my model. When an ID is used that doesn't exist, at latest the query execution (like `connect` or the `where` in an `update`) will fail (because this does not exists as a foreign key). This is validation enough I think.

I'm happy with my current setup, I just still feel that I haven't thought this through completely.

Thanks again!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
